src/Database/Traits/SoftDelete.php

```diff
@@ -62,6 +75,46 @@ public static function bootSoftDelete()
              */
             return $model->fireEvent('model.afterRestore', halt: true);
         });
+
+        foreach ([
+            AttachMany::class,
+            AttachOne::class,
+            BelongsToMany::class,
+            HasMany::class,
+            HasOne::class,
+            MorphMany::class,
+            MorphOne::class,
+            MorphToMany::class,
+        ] as $relationClass) {
+            $relationClass::extend(function () {
+                // Prevent double-defining the dynamically added properties and methods below
+                if ($this->methodExists('softDeletable')) {
+                    return;
+                }
```

Comment:
This still concerns me as I think it becomes an n+1 issue where if a model uses SoftDelete then every instance of that model will cause an additional function to be added to these relation classes, even if that function exits early. Thoughts @bennothommo @jaxwilko?

Reply:
It would be minimised because it would be only on each "class" that uses it, not each model instance. But otherwise, I'm running out of ideas. I tried doing a "local" extension but I think that's too late in the boot process for it to be available in relation methods.